File: packages/helm-runtime-sdk/node/test/smoke.mjs
// Smoke test of the Node client against a running daemon.
//
// Run by test/conformance (TestPythonAndNodeClientsSmoke) with HELM_API,
// HELM_TOKEN and HELM_STAGE_DIR set for a studio holding kv, records, assets
// and gallery. Exits non-zero, naming the failed check, on any mismatch.

import { writeFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { fromEnv, HelmError } from "../src/index.js";

function check(cond, what) {
  if (!cond) {
    console.log("FAIL: " + what);
    process.exit(1);
  }
}

async function refused(promise) {
  try {
    await promise;
  } catch (e) {
    return e;
  }
  return null;
}

const helm = fromEnv();
const me = await helm.me.get();
check(me.studio_id === process.env.SMOKE_STUDIO, "me.studio_id");

const doc = await helm.kv.put("ui", "node", { n: 1 });
let e = await refused(helm.kv.put("ui", "node", { n: 2 }, { ifMatch: "stale" }));
check(e instanceof HelmError && e.kind === "Conflict" && e.code === "etag_mismatch", "etag conflict kind");
const patched = await helm.kv.patch("ui", "node", { m: 2 }, { ifMatch: doc.etag });
check(patched.doc.n === 1 && patched.doc.m === 2, "merge patch");

const rec = await helm.records.insert("takes", { seed: 7 });
const page = await helm.records.query("takes", { where: ["seed:eq:7"] });
check(page.items.length === 1 && page.items[0].id === rec.id, "records filter");

const staged = join(process.env.HELM_STAGE_DIR, "node-take.mp4");
writeFileSync(staged, "node take");
const asset = await helm.assets.adopt({ path: staged, kind: "video" });
check(!existsSync(staged), "adopt unlinked the stage entry");
const part = await helm.assets.read(asset.id, { range: "bytes=0-3" });
check(part.status === 206 && (await part.text()) === "node", "range read");

const item = await helm.gallery.add({ kind: "video", asset_id: asset.id, params: { seed: 7 }, tags: ["node"] });
const found = await helm.gallery.query({ tag: ["node"] });
check(found.items.length === 1 && found.items[0].id === item.id, "gallery tag query");

e = await refused(helm.jobs.list());
check(e instanceof HelmError && e.kind === "Forbidden" && e.details.capability === "jobs", "capability_required details");

console.log("ok");

File: packages/helm-runtime-sdk/python/tests/smoke.py
"""Smoke test of the Python client against a running daemon.

Run by test/conformance (TestPythonAndNodeClientsSmoke) with HELM_API,
HELM_TOKEN and HELM_STAGE_DIR set for a studio holding kv, records, assets and
gallery. Exits non-zero, naming the failed check, on any mismatch.
"""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from helm_runtime_sdk import HelmError, from_env  # noqa: E402


def check(cond, what):
    if not cond:
        print("FAIL: " + what)
        sys.exit(1)


helm = from_env()
me = helm.me.get()
check(me["studio_id"] == os.environ["SMOKE_STUDIO"], "me.studio_id")

doc = helm.kv.put("ui", "py", {"n": 1})
try:
    helm.kv.put("ui", "py", {"n": 2}, if_match="stale")
    check(False, "stale If-Match was accepted")
except HelmError as e:
    check(e.kind == "Conflict" and e.code == "etag_mismatch", "etag conflict kind: %s %s" % (e.kind, e.code))
patched = helm.kv.patch("ui", "py", {"m": 2}, if_match=doc["etag"])
check(patched["doc"] == {"n": 1, "m": 2}, "merge patch")

rec = helm.records.insert("takes", {"seed": 42, "prompt": "a cafe"})
page = helm.records.query("takes", where=["seed:eq:42"], limit=10)
check([r["id"] for r in page["items"]] == [rec["id"]], "records filter")

staged = os.path.join(os.environ["HELM_STAGE_DIR"], "py-take.mp4")
with open(staged, "wb") as f:
    f.write(b"python take")
asset = helm.assets.adopt({"path": staged, "kind": "video"})
check(not os.path.exists(staged), "adopt unlinked the stage entry")
body = helm.assets.read(asset["id"], range="bytes=0-5").read()
check(body == b"python", "range read: %r" % body)
up = helm.assets.upload(b"py upload", "application/octet-stream", kind="other", filename="u.bin")
check(up["bytes"] == 9, "upload")

item = helm.gallery.add({"kind": "video", "asset_id": asset["id"], "params": {"seed": 42}, "tags": ["py"]})
found = helm.gallery.query(tag=["py"])
check([i["id"] for i in found["items"]] == [item["id"]], "gallery tag query")

try:
    helm.jobs.list()
    check(False, "jobs without the capability")
except HelmError as e:
    check(e.kind == "Forbidden" and e.details.get("capability") == "jobs", "capability_required details")

print("ok")
